Widget support

kshala-ford opened the pull request:
This proposal is about widget support for SDL applications on modern infotainment systems and large displays. At the same time the proposed feature should prepare SDL for multi screen and multi display support.

kshala-ford commented:
@jordynmackool The PR is ready for review.

jordynmackool commented on proposals/NNNN-widget-support.md:
```diff
@@ -0,0 +1,248 @@
+
+### Screen templates
+
+Today SDL comes with a set of templates available the app can choose from. The information what templates are available to the app is provided in 
```
Update to:
**Today, SDL comes with a set of templates available for the app developer to choose from. The information for what templates are available for the apps are provided in**

jordynmackool commented on proposals/NNNN-widget-support.md:
```diff
@@ -0,0 +1,248 @@
+- text fields, the name, width etc.
+- image fields e.g. graphic, or soft button image resolutions
+
+The template design comes from the head unit. However predefined template names are defined in `PredefinedLayout` enum. The head unit should provide template designs for the predefined template names. Still it can also provide additional custom templates with self defined names. 
```
Update to: 
The template design comes from the head unit. However predefined template names are defined in `PredefinedLayout` enum. The head unit should provide template designs for the predefined template names. **Still,** it can also provide additional custom templates with self defined names.

jordynmackool commented on proposals/NNNN-widget-support.md:
```diff
@@ -0,0 +1,248 @@
+
+### Screen types supported
+
+In order to inform the app what screen types are supported the struct `DisplayCapabilities` should be extended with a single optional parameter called `screenTypesSupported`.
```
Update to:
In order to inform the app what screen types are **supported,** the struct `DisplayCapabilities` should be extended with a single optional parameter called `screenTypesSupported`.

jordynmackool commented on proposals/NNNN-widget-support.md:
````diff
@@ -0,0 +1,248 @@
+</struct>
+```
+
+This parameter does not inform the app how many screens of the type are supported but it informs if a display supports widgets or a main app screen. The HMI decides on 
````
Should `the` be `this`?

This parameter does not inform the app how many screens of **the** type are supported but it informs if a display supports widgets or a main app screen. The HMI decides on

kshala-ford replied:
changed it to *per*

jordynmackool commented on proposals/NNNN-widget-support.md:
```diff
@@ -0,0 +1,248 @@
+
+## Potential downsides
+
+With current design displays and screens are not clearly separated. Together with multi-display support there will be a proposal to redesign display and screen capabilities. This proposal should still be valid as it focuses on mutli screen only.
```
Update to:
**Current** design displays and screens are not clearly separated. Together with multi-display support there will be a proposal to redesign display and screen capabilities. This proposal should still be valid as it focuses on **multi** screen only.

jordynmackool commented on proposals/NNNN-widget-support.md:
```diff
@@ -0,0 +1,248 @@
+
+## Impact on existing code
+
+On SDL there are additional RPCs to be added. The HMI needs to design templates for widgets. This proposal did not focus on the mobile changes. For mobile libraries it's necessary to add the new RPCs and add support for multiple screen managers. This will also be part of a separate proposal.
```
Change to:
On SDL there are additional RPCs **that need to be added.** The HMI needs to design templates for widgets. This proposal did not focus on the mobile changes. For mobile libraries it's necessary to add the new RPCs and add support for multiple screen managers. This will also be part of a separate proposal.

jordynmackool commented on proposals/NNNN-widget-support.md:
```diff
@@ -0,0 +1,248 @@
+
+The widget screen is a small screen type to provide quick information or shortcut softbuttons. Widgets don't follow HMI levels as main screen. Once a widget is created the all RPCs addressed to this widget follow policies based on HMI_LIMITED. This policy rule is due to the nature of widgets as they can be always avaiable to the user or should be ready to use when accessing the widget area (see a ...
+
+Still widgets have affect to the HMI level of the app's main screen. Depending on the app policies:
```
Change to:
Still widgets have **effects** to the HMI level of the app's main screen. Depending on the app policies:

kshala-ford commented on proposals/NNNN-widget-support.md:
```diff
@@ -0,0 +1,248 @@
+
+### Apps, Displays, Screens
+
+Today when we speak of SDL apps once they register they get a default screen allocated on the IVI head unit. Strictly speaking the Show RPC is manipulating the content presented in the single default screen. For media apps some IVI reuse the content on clusters or on IVI overview screens (SYNC3 home screen). Adding the capability for apps to address a specific screen per Show request allows apps  ...
```
@jordynmackool Do you agree with the following revision? 
> As of today, once an SDL app registers it gets a default screen allocated on the IVI head unit.

kshala-ford commented:
@jordynmackool I updated the proposal and fixed all the grammars.

kshala-ford commented:
@jordynmackool ping

jordynmackool commented on proposals/NNNN-widget-support.md:
````diff
@@ -0,0 +1,248 @@
+</function>    
+```
+
+The additional parameter can used by the system to tell the app the HMI level of any screen instance. The parameter should be omitted for the predefined default screen for better backward compatibility. For other screens the screen ID specified by the app should be set in this parameter.
````
Update to:
The additional parameter can **be** used by the system to tell the app the HMI level of any screen instance. The parameter should be omitted for the predefined default screen for better backward compatibility. For other screens the screen ID specified by the app should be set in this parameter.

jordynmackool commented on proposals/NNNN-widget-support.md:
```diff
@@ -0,0 +1,248 @@
+
+The additional parameter can used by the system to tell the app the HMI level of any screen instance. The parameter should be omitted for the predefined default screen for better backward compatibility. For other screens the screen ID specified by the app should be set in this parameter.
+
+By default widgets created are in HMI_BACKGROUND. This allows the app to fill the widget with content after it's created. If they get visible on the display the HMI should notify the app that the widget is now in HMI_FULL. This can be the case if the driver selects the widget area (e.g. the home screen shows app widgets). As widgets are not launchable the should automatically transit to HMI_FULL  ...
```
Update to:
**Default** widgets **are** created in HMI_BACKGROUND. This allows the app to fill the widget with content after it's created. If they get visible on the **display, the HMI** should notify the app that the widget is now in HMI_FULL. This can be the case if the driver selects the widget area (e.g. the home screen shows app widgets). As widgets are not launchable **they** should automatically transit to HMI_FULL once they are visible.

jordynmackool commented on proposals/NNNN-widget-support.md:
```diff
@@ -0,0 +1,248 @@
+
+By default widgets created are in HMI_BACKGROUND. This allows the app to fill the widget with content after it's created. If they get visible on the display the HMI should notify the app that the widget is now in HMI_FULL. This can be the case if the driver selects the widget area (e.g. the home screen shows app widgets). As widgets are not launchable the should automatically transit to HMI_FULL  ...
+
+To limit the changes in the RPC and support backward compatibility audioStreamingState should be provided to all screens of one app. The HMI and SDL core must make sure to not send different audio states per app.
```
Update to:
To limit the changes in the RPC and support backward **compatibility,** audioStreamingState should be provided to all screens of one app. The HMI and SDL core must make sure to not send different audio states per app.

jordynmackool commented on proposals/NNNN-widget-support.md:
```diff
@@ -0,0 +1,248 @@
+
+#### Screen duplication by SDL libraries
+
+To reduce complexity on the head unit screen duplication can be supported by the SDL libraries. This could allow more flexibility to enable or display duplication for existing screens. However this increases the number of RPCs to be sent by the app and the user might see a delay in screen update (duplication not synchronized).
```
Update to:
To reduce complexity on the head unit **screen, duplication** can be supported by the SDL libraries. This could allow more flexibility to enable or display duplication for existing screens. However this increases the number of RPCs to be sent by the app and the user might see a delay in screen update (duplication not synchronized).

kshala-ford commented on proposals/NNNN-widget-support.md:
```diff
@@ -0,0 +1,248 @@
+<enum "PredefinedScreens" since="5.1">
+<element name="DEFAULT_SCREEN" value="0">
+  <description>
+    The default screen is a main screen precreated on behalfs of the app.
```
I don't think this explains it better. I'll add ad dash instead.

kshala-ford commented:
> In the future, please don't make changes to the PR after asking for a review. I have re- reviewed.

I'm sorry I found a type and thought it doesn't hurt to fix it. Won't happen again. I addressed all issues and hope the document now meets the quality required for review.
